lib/OpenQA/Worker/Cache.pm

```diff
@@ -35,173 +37,325 @@ our (@ISA, @EXPORT);
+sub deploy_db {
+    local $/;
+    my $sql = <DATA>;
+    log_info "Deploying DB: $sql";
+    $dbh = DBI->connect($dsn, undef, undef, {RaiseError => 1, PrintError => 1, AutoCommit => 0})
+      or die("Could not connect to the dbfile.");
+    $dbh->do($sql);
+    $dbh->commit;
+    $dbh->disconnect;
 }
 
 sub init {
-    my $class;
     ($host, $location) = @_;
     $db_file = catdir($location, 'cache.db');
-    @{$cache->{$host}} = read_db();
-    log_debug(__PACKAGE__ . ": Initialized with $host at $location");
+    $dsn = "dbi:SQLite:dbname=$db_file";
+    deploy_db unless (-e $db_file);
+    $dbh = DBI->connect($dsn, undef, undef, {RaiseError => 1, PrintError => 1, AutoCommit => 1})
+      or die("Could not connect to the dbfile.");
+    $cache_real_size = 0;
+    cache_cleanup();
+    #Ideally we only need $limit, and $need no extra space
+    check_limits(0);
+    log_info(__PACKAGE__ . ": Initialized with $host at $location, current size is $cache_real_size");
 }
 
 sub download_asset {
-    my ($id, $type, $asset) = @_;
+    my ($id, $type, $asset, $etag) = @_;
 
     open(my $log, '>>', "autoinst-log.txt") or die("Cannot open autoinst-log.txt");
     local $| = 1;
-    print $log "CACHE: Locking $asset\n";
-
-    print $log "Attemping to download: $host $asset, $type, $id\n";
     my $ua = Mojo::UserAgent->new(max_redirects => 2);
     $ua->max_response_size(0);
-    my $tx = $ua->build_tx(GET => sprintf '%s/tests/%d/asset/%s/%s', $host, $id, $type, basename($asset));
+    my $url = sprintf '%s/tests/%d/asset/%s/%s', $host, $id, $type, basename($asset);
+    print $log "Downloading " . basename($asset) . " from $url\n";
+    my $tx = $ua->build_tx(GET => $url);
+    my $headers;
 
     $ua->on(
         start => sub {
             my ($ua, $tx) = @_;
             my $progress     = 0;
             my $last_updated = time;
+            $tx->req->headers->header('If-None-Match' => qq{$etag}) if $etag;
             $tx->res->on(
                 progress => sub {
                     my $msg = shift;
+                    $msg->finish if $msg->code == 304;
                     return unless my $len = $msg->headers->content_length;
+
                     my $size = $msg->content->progress;
+                    $headers = $msg->headers if !$headers;
                     my $current = int($size / ($len / 100));
-                    local $| = 1;
                     # Don't spam the webui, update only every 5 seconds
                     if (time - $last_updated > 5) {
                         update_setup_status;
+                        toggle_asset_lock($asset, 1);
                         $last_updated = time;
                         if ($progress < $current) {
                             $progress = $current;
-                            print $log "Downloading $asset :", $size == $len ? 100 : $progress, "\n";
+                            print $log "CACHE: Downloading $asset :", $size == $len ? 100 : $progress . "\n";
```

Comment:
To me `93% done` sounds a bit more confusing. Also, once the asset is downloaded, an actual message saying that it has been downloaded will show up.